src/sys/system_v/shm.rs

```diff
@@ -0,0 +1,476 @@
+    ///         MY_KEY,
+    ///         ShmgetFlag::empty(),
+    ///         Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+    ///     )?;
+    /// }
+    /// # Ok::<(), Errno>(())
+    /// ```
+    ///
+    /// [`shmget(2)`]: https://man7.org/linux/man-pages/man2/shmget.2.html
+    pub unsafe fn shmget(
+        key: key_t,
+        shmget_flag: ShmgetFlag,
+        mode: Mode,
+    ) -> Result<Self> {
+        let size = std::mem::size_of::<T>();
+        let flags = mode.bits() as i32 | shmget_flag.bits();
+        let id = Errno::result(unsafe { libc::shmget(key, size, flags) })?;
+        Ok(Self {
+            id,
+            _phantom: PhantomData,
+        })
+    }
+
+    // -- Private --
+
+    /// Attaches the System V shared memory segment identified by `shmid` to the
+    /// address space of the calling process.
+    ///
+    /// This is called automatically on [`Shm::attach`].
+    ///
+    /// For more information, see [`shmat(2)`].
+    ///
+    /// [`shmat(2)`]: https://man7.org/linux/man-pages/man2/shmat.2.html
+    fn shmat(&self, shmat_flag: ShmatFlag) -> Result<*mut T> {
+        Errno::result(unsafe {
+            libc::shmat(self.id, ptr::null(), shmat_flag.bits())
+        })
+        .map(|ok| ok.cast::<T>())
+    }
+}
+
+#[derive(Debug)]
+/// Safe wrapper around a SystemV shared memory segment data
+///
+/// This is a smart pointer, and so implement the [`Deref`] and [`DerefMut`] traits.
+/// This means that you can work with the shared memory segment like you would with a [`Box`].
+///
+/// This type does not automatically destroy the shared memory segment, but
+/// only detach from it using RAII.
+///
+/// To delete a shared memory segment, use [`SharedMemory::shmctl`], with the key [`ShmctlFlag::IPC_RMID`].
+///
+/// # Example
+///
+/// ```no_run
+/// # use nix::errno::Errno;
+/// # use nix::sys::system_v::shm::*;
+/// # use nix::sys::stat::Mode;
+/// #
+/// struct MyData(i64);
+/// const MY_KEY: i32 = 1337;
+///
+/// let mem_segment = Shm::<MyData>::create_and_connect(
+///     MY_KEY,
+///     Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+/// )?;
+/// let mut shared_memory = mem_segment.attach(ShmatFlag::empty())?;
+///
+/// // This is writing to the stored [`MyData`] struct
+/// shared_memory.0 = 0xDEADBEEF;
+/// # Ok::<(), Errno>(())
+/// ```
+///
+pub struct SharedMemory<T> {
+    id: c_int,
+    shm: ManuallyDrop<Box<T>>,
+}
+
+impl<T> Deref for SharedMemory<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.shm
+    }
+}
+impl<T> DerefMut for SharedMemory<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.shm
+    }
+}
+
+impl<T> Drop for SharedMemory<T> {
+    fn drop(&mut self) {
+        Self::shmdt(self).expect("SharedMemory detach from SysV IPC");
+    }
+}
+
+impl<T> SharedMemory<T> {
+    /// Performs control operation specified by `cmd` on the System V shared
+    /// memory segment given by `shmid`.
```

Comment:
This `shmid` identifier is not a parameter but stored in `SharedMemory`

Reply:
Fixed